src/OpenTelemetry/Metrics/MeterProviderBuilderSdk.cs

```diff
@@ -31,7 +31,7 @@ internal MeterProviderBuilderSdk()
 
         internal List<MeasurementProcessor> MeasurementProcessors { get; } = new List<MeasurementProcessor>();
 
-        internal List<MetricProcessor> ExportProcessors { get; } = new List<MetricProcessor>();
+        internal List<KeyValuePair<MetricProcessor, int>> ExportProcessors { get; } = new List<KeyValuePair<MetricProcessor, int>>();
```

Comment:
Not a fully baked thought...

Would it make sense for metric export processors to derive from [`BaseExportProcessor`](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/BaseExportProcessor.cs)? `MetricProcessor` currently derives from `BaseProcessor<MetricItem>`.

Would something like this work instead of having a list of `KeyValuePair` here?

```csharp
public class MetricExportProcessor : BaseExportProcessor<MetricItem>
{
    public int CollectionInterval { get; }
}
```

Reply:
Yes, I know what you are saying, and it does make sense, and it is a viable option.

The key (I think) is who owns/manage the collection interval?

In your example, the logical conclusion is to embed the CollectionInterval into the instance of the Processor. So, this will necessitate a new instance of a Processor per collection interval. Plus, the coordination of the same interval with the collection process (callback observables) itself.

Currently, my viewpoint is, I think, the CollectionInterval has nothing to do with the Processor itself. It's collection which deals with coordinating multiple steps. Thus, it's stored outside of the Processor itself.

Separately...
We can clean up how the KeyValuePair<MetricProcessor, int> syntax look by declaring a new struct (or SystemTuple, but not supported easily on all versions) if that helps readability.

Reply:
Also, as a TODO item, we should consolidate on the same BaseProcessor<T> or BaseExportProcessor<T>. Currently, it is designed around Trace, and I didn't want to spend too much time there.

Reply:
> Currently, my viewpoint is, I think, the CollectionInterval has nothing to do with the Processor itself. It's collection which deals with coordinating multiple steps. Thus, it's stored outside of the Processor itself.

I'm probably still in the process of learning the complexities here - I was pondering whether the exporter is kind of like a batching processor where the collection interval defines how often metrics are flushed.

Though the thought of defining this as `Dictionary<interval, List<Stuff>>` suggests that the SDK itself would be doing a bunch of `Stuff` on some interval - so if this is the case, then it makes sense - but what are some examples of the other "stuff" besides exporting?

> Also, as a TODO item, we should consolidate on the same BaseProcessor or BaseExportProcessor. Currently, it is designed around Trace, and I didn't want to spend too much time there.

Yea, I noted the trace-centric design here too. Totally makes sense to defer thoughts of refactoring here to future PRs.

Reply:
IMHO...

I struggle with the fact that a metric (aka OTLP metric) is a pre-aggregation for a given time period. To simplify things, I tied the Collection interval (which does aggregation) together with Export interval (which defines the time period). So, if we have another Exporter with a different interval, we will need another set of aggregators specifically for that interval.

A different approach may be...
We do collection (aka aggregation) as a fast interval. And then, the exporter can collect at a slower interval, but will have to merge the metrics (combine multiple fast interval metrics into one metric that covers the full time period, aka aggregation!) before sending out. If we adopt this approach, then, the solution looks like a degenerative version of the first solution.  It's just sending "raw data" (cause it's very fast interval) to the exporter for aggregation (aka merge). BTW, this work needs to be done by Collector regardless, but I'm hoping to keep the complexity of the "merge" logic out of the SDK.

As for the other `Stuff`, they are Metric Processors. For now, I think filtering and manipulation of a pre-aggregated metric in first phase. Future phases may include merging metrics (LOL).  Another near-future is support for "pull" exporters, which may need a Metric Processor that store `state` based on pre-aggregate metrics.